A29-xds-tls-security.md

```diff
@@ -0,0 +1,446 @@
+
+### Certificate Provider Plugin Framework
+
+The pluggable Certificate Provider framework in gRPC underlies the use of `CertificateProviderInstance`
+configuration. The framework supports "Certificate Providers" as pluggable and custom components (plugins)
+used for dynamically fetching certificates used during the TLS handshake of gRPC channel and server connections.
+
+This document discusses just enough of the framework to explain the use-case. Users who want to
+implement new plugins will need to understand the detailed design of the framework in the
+respective language.
+
+Envoy was the original and only consumer of the xDS protocol and its security features were tied to
+Envoy's security capabilities such as fetching certificates from the file-system or
+using a separate SDS protocol to fetch certificates from an agent (that's not defined in the xDS
+architecture). Instead of emulating Envoy's functionality and limitations in this area, we decided to
+have a generic Certificate Provider plugin architecture in gRPC with the following characteristics:
+
+* the xDS control plane is not aware of the plugin specifics including its name, instantiation
+mechanism or configuration requirements.
+
+* all of these are abstracted out into a single “plugin-instance-name” which is resolved into an
+actual Certificate provider instance using the local bootstrap file and the framework (as explained
+later).
+
+* the xDS control plane only references the “plugin-instance-name”s: it uses one for the local
+certificates and another one for root certificates to validate the peer certificates (aka
+[`CertificateValidationContext`][CVC-proto]).
+
+[CVC-proto]: https://github.com/envoyproxy/envoy/blob/c94e646e0280e4c521f8e613f1ae2a02b274dbbf/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L236
+
+#### Main Elements of the Framework
+
+**`CertificateProvider`** represents an actual plugin that fetches or "mints" the required certificates
+and keys. One example of a (already implemented) plugin is a `FileWatcherCertificateProvider` which
+monitors certain files in the file system and reads those files on updates to get the latest
+certificates and keys and provide those to the consumers after converting them to the canonical
+format e.g. `java.security.PrivateKey` and `java.security.cert.X509Certificate` in Java.
+Another example is a plugin that periodically "mints" new certificates by creating
+CSRs at regular intervals and getting those signed by a Certification Authority (CA) that is
+configured into the plugin.
+
+A client registers itself as a **`Watcher`**  of the plugin to receive certificate updates.
+The plugin supports multiple **`Watchers`** to be registered and uses a **`DistributorWatcher`**
+internally to propagate a single update to multiple **`Watchers`**. A plugin caches the latest
+certificate (and key) it has fetched (or minted) and delivers them to every newly registered
+watcher.
+
```

Comment:
We probably need to add a line saying something to the effect:

Go uses a fetch-style API instead of a watch-style API, where the component using the certificates and keys calls into the `CertificateProvider` whenever it needs them. The high-level concepts are identical except for the API style.